File: mobile-ios/Modules/Domain/Sources/Repositories/CatalogRepository.swift
import Foundation

public enum ProductSort: String, Sendable, CaseIterable {
    case newest
    case priceAscending = "price_asc"
    case priceDescending = "price_desc"
}

public struct ProductQuery: Sendable, Equatable {
    public var page: Int
    public var pageSize: Int
    public var categoryID: String?
    public var search: String?
    public var sort: ProductSort
    public var onlyDiscounted: Bool
    public var onlyNew: Bool
    public var inStock: Bool

    public init(
        page: Int = 1,
        pageSize: Int = 20,
        categoryID: String? = nil,
        search: String? = nil,
        sort: ProductSort = .newest,
        onlyDiscounted: Bool = false,
        onlyNew: Bool = false,
        inStock: Bool = false
    ) {
        self.page = page
        self.pageSize = pageSize
        self.categoryID = categoryID
        self.search = search
        self.sort = sort
        self.onlyDiscounted = onlyDiscounted
        self.onlyNew = onlyNew
        self.inStock = inStock
    }
}

public protocol CatalogRepository: Sendable {
    func products(_ query: ProductQuery, language: String) async throws -> Page<Product>
    func product(id: String, language: String) async throws -> Product
    func similarProducts(id: String, language: String) async throws -> [Product]
    func categories(language: String) async throws -> [ProductCategory]
}

File: mobile-android/app/src/main/java/com/erenlermarket/app/domain/repository/CatalogRepository.kt
package com.erenlermarket.app.domain.repository

import com.erenlermarket.app.domain.model.Page
import com.erenlermarket.app.domain.model.Product
import com.erenlermarket.app.domain.model.ProductCategory
import com.erenlermarket.app.domain.model.ProductQuery

interface CatalogRepository {
    suspend fun products(query: ProductQuery): Page<Product>
    suspend fun product(id: String): Product
    suspend fun similarProducts(id: String): List<Product>
    suspend fun categories(): List<ProductCategory>
}
